rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,208 @@
+
+It's worth noting that this change will lead to slower load times for browsers that only support HTTP/1.1 - even when considering the reduction in page size. ([More information on HTTP/1.1's limitations][http_limitations]). This will impact Internet Explorer 11 on Windows 8 and below; the first versions of Chrome, Firefox, and Opera that supported HTTP/2 were released in early 2015.
+
+This balance between making the site load faster for a majority of users is not just a numbers game - the users who are most likely to be using older browsers are those that are likely to least afford the upgrade cost for a newer device, and they're likely to be the users who most need the information on GOV.UK.
+
+Whilst this proposed change will make things slower for users whose browser doesn't support HTTP/2 or HTTP/3, it won't prevent pages from loading. This proposed change should be accepted because it will dramatically improve page performance for almost all visitors without blocking access for a small number of visitors.
+
+### Non-rendering applications that use the components gem
+
+There are applications that use the components gem, but don't render public pages on GOV.UK - for example, publishing applications and the Govspeak gem. These may have a different set up with different ways of rendering pages.
+
+These changes should be optional - even if this is the default behaviour, it should be able to be turned off at a component by component level and at an application level. This could be done with a helper method that checks both if the component has already been used, and if adding the component assets are allowed:
+
+```ruby
+# component_asset_helper.rb - pseudo code example:
+
+def component_assets_allowed (component_name)
+  # check:
+  #  - if the component has already had the assets added to the page
+  # - if the component assets loading method is allowed at an application level
+  # - if the individual component has used an option to turn individual assets on or off
+  @components_in_use.include?(component_name) || ENV.component_asset_loading !== "bulk" || use_individual_assets === true
+end
+```
+
+## Other solutions that have been considered and rejected
+
+* Put all components used across GOV.UK in one stylesheet and one JavaScript file - this increases page weight and doesn't solve cache invalidation
+* Serve one stylesheet and one JavaScript file for all of GOV.UK - this increases page weight and doesn't solve cache invalidation
+* Serve one stylesheet and one JavaScript file per template - this doesn't solve cache invalidation when one component changes
+* Serve all component assets used in an application in individual files - this increases page weight for single page visitors, though often decreased it for people who visited more than one page
```

Comment:
This is about serving all individual assets for an application, rather than for the page being viewed - so a page rendered by collections would include all of the assets needed for the entire app, rather than just that page. Kind of like the current behaviour, except with the assets loaded individually instead of being concatenated.